### ISSUE TITLE
Feature request: add macro @produced for simulation parameters that depend on other parameters.

### ISSUE THREAD

**Issue by aramirezreyes** [JuliaDynamics/DrWatson.jl#358]
**Is your feature request related to a problem? Please describe.**
This feature request concerns the creation of dictionaries of parameters. From the documentation example:
```julia
using DrWatson
general_args = Dict(
    "model" => "barkley",
    "noise" => [0.075, 0.050, 0.025],
    "noisy_training" => true,
    "N" => 100,
)
dicts = dict_list(general_args)
```
Let's say we want to have another parameter, noise2, that for each simulation is double that of "noise" apart from another value.

Currently I would need to do:

```julia
map(dicts) do dict
    dict["noise2"] = 2*dict["noise"]
end
```

But this has the disadvantage of:

1. more verbose and
2.  the original `general_args` is now not enough to reproduce all the simulations.
3. It does not work if the we also want to have another value in all the simulations (see example below).

**Describe the solution you'd like**
It would be nice to have a `@produced` (name suggested by @Datseris) that allows to do something like:

```julia
using DrWatson
general_args = Dict(
    "model" => "barkley",
    "noise" => [0.075, 0.050, 0.025],
    "noise2" => [1.0, @produced("noise" -> 2*"noise")]
    "noisy_training" => true,
    "N" => 100,
)
dicts = dict_list(general_args)
```

This should create 6 different Dicts of simulation parameters:
```julia
julia> dicts = dict_list(general_args)
6-element Vector{Dict{String, Any}}:
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.075, "noisy_training" => true)
 Dict("noise2" => 0.15, "model" => "barkley", "N" => 100, "noise" => 0.075, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.05, "noisy_training" => true)
 Dict("noise2" => 0.1, "model" => "barkley", "N" => 100, "noise" => 0.05, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.025, "noisy_training" => true)
 Dict("noise2" => 0.05, "model" => "barkley", "N" => 100, "noise" => 0.025, "noisy_training" => true)
 ```

**Describe alternatives you've considered**

Another option could be using `@onlyif` in the following fashion:
```julia
using DrWatson
general_args = Dict(
    "model" => "barkley",
    :noise => [0.075, 0.050, 0.025],
    "noise2" => [1.0, @onlyif("noise" == 0.075,0.15), @onlyif("noise" == 0.050, 0.1), @onlyif("noise" == 0.025, 0.05)],
    "noisy_training" => true,
    "N" => 100,
)
dicts = dict_list(general_args)
```

This solves the second and third pain points but not the first one.

Would the community consider this to be a desirable feature?

**Comment by Datseris** [comment]
I don't get this:
```julia
    "noise2" => [1.0, @produced(:noise -> 2*:noise)]
```
So what would the value of `noise2` be? 1 or 2*noise? It's ambiguous here with that rogue `1`. You said that

> This should create 6 different Dicts of simulation parameters.

What are these exactly? 3 versions with noise2 being 1, and 3 versions with noise2 being 2*noise?

**Comment by aramirezreyes** [comment]
> What are these exactly? 3 versions with noise2 being 1, and 3 versions with noise2 being 2*noise?

Yes, so the result is:

```julia
julia> dicts = dict_list(general_args)
6-element Vector{Dict{Any, Any}}:
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, :noise => 0.075, "noisy_training" => true)
 Dict("noise2" => 0.15, "model" => "barkley", "N" => 100, :noise => 0.075, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, :noise => 0.05, "noisy_training" => true)
 Dict("noise2" => 0.1, "model" => "barkley", "N" => 100, :noise => 0.05, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, :noise => 0.025, "noisy_training" => true)
 Dict("noise2" => 0.05, "model" => "barkley", "N" => 100, :noise => 0.025, "noisy_training" => true)
```

This would be consistent with how `@onlyif` can be mixed with other unconstrained parameters in the same "rogue" fashion.

**Comment by aramirezreyes** [comment]
(added this last expected output to the original post).

**Comment by aramirezreyes** [comment]
I created a prototype that so far works like this:

```julia
julia> using DrWatson

julia> general_args = Dict(
           "model" => "barkley",
           "noise" => [0.075, 0.050, 0.025],
           "noise2" => [1.0, ComputedParameter("noise", x -> 2x)],
           "noisy_training" => true,
           "N" => 100,
       )
       dicts = dict_list(general_args)
6-element Vector{Dict{String, Any}}:
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.075, "noisy_training" => true)
 Dict("noise2" => 0.15, "model" => "barkley", "N" => 100, "noise" => 0.075, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.05, "noisy_training" => true)
 Dict("noise2" => 0.1, "model" => "barkley", "N" => 100, "noise" => 0.05, "noisy_training" => true)
 Dict("noise2" => 1.0, "model" => "barkley", "N" => 100, "noise" => 0.025, "noisy_training" => true)
 Dict("noise2" => 0.05, "model" => "barkley", "N" => 100, "noise" => 0.025, "noisy_training" => true)
```

Any comments are appreciated.

**Comment by aramirezreyes** [comment]
Closed by #359!
